File: zatca/zatca/qr.py
"""Assemble the ZATCA invoice QR payload.

This is what gets printed on the customer's receipt. It is produced on the
device, at the moment of sale, with no network — which is the whole reason each
tablet holds its own CSID.

Money arrives here as integer halalas (as it is stored everywhere else) and is
formatted to two decimals only at this boundary, because ZATCA wants a decimal
string. The conversion is integer division, never float arithmetic.
"""

from __future__ import annotations

import datetime as dt
from dataclasses import dataclass

from . import tlv
from .signing import sign_digest


def halalas_to_decimal_string(halalas: int) -> str:
    """1234 -> '12.34'. Integer maths only; no float ever touches a total."""
    if halalas < 0:
        raise ValueError("amount cannot be negative")
    return f"{halalas // 100}.{halalas % 100:02d}"


def zatca_timestamp(when: dt.datetime) -> str:
    """ISO 8601 in UTC with a trailing Z, no microseconds."""
    if when.tzinfo is None:
        raise ValueError(
            "timestamp must be timezone-aware — a naive local time on a device "
            "with drifting clock produces invoices ZATCA rejects"
        )
    return when.astimezone(dt.timezone.utc).replace(microsecond=0).isoformat().replace(
        "+00:00", "Z"
    )


@dataclass(frozen=True)
class QrInput:
    seller_name: str          # Arabic name as registered with ZATCA
    vat_number: str           # 15 digits
    issued_at: dt.datetime
    total_with_vat: int       # halalas
    vat_total: int            # halalas
    invoice_hash: str         # base64
    public_key_der: bytes
    csid_signature: str       # base64, ZATCA's signature over the public key


def build_qr(data: QrInput, private_pem: bytes) -> str:
    """Return the base64 TLV payload to encode as a QR on the receipt."""
    if len(data.vat_number) != 15 or not data.vat_number.isdigit():
        raise ValueError("VAT registration number must be 15 digits")
    if data.vat_total > data.total_with_vat:
        raise ValueError("VAT cannot exceed the invoice total")

    # The stamp is over the invoice hash, which is what binds the QR to the
    # document it was printed for.
    signature = sign_digest(private_pem, data.invoice_hash.encode("ascii"))

    return tlv.encode_base64({
        tlv.Tag.SELLER_NAME:    data.seller_name,
        tlv.Tag.VAT_NUMBER:     data.vat_number,
        tlv.Tag.TIMESTAMP:      zatca_timestamp(data.issued_at),
        tlv.Tag.INVOICE_TOTAL:  halalas_to_decimal_string(data.total_with_vat),
        tlv.Tag.VAT_TOTAL:      halalas_to_decimal_string(data.vat_total),
        tlv.Tag.INVOICE_HASH:   data.invoice_hash,
        tlv.Tag.SIGNATURE:      signature,
        tlv.Tag.PUBLIC_KEY:     data.public_key_der,
        tlv.Tag.CSID_SIGNATURE: data.csid_signature,
    })


def inspect(qr_base64: str) -> dict[str, str]:
    """Decode a QR payload for support and debugging."""
    raw = tlv.decode_base64(qr_base64)
    names = {
        tlv.Tag.SELLER_NAME: "seller_name",
        tlv.Tag.VAT_NUMBER: "vat_number",
        tlv.Tag.TIMESTAMP: "timestamp",
        tlv.Tag.INVOICE_TOTAL: "total_with_vat",
        tlv.Tag.VAT_TOTAL: "vat_total",
        tlv.Tag.INVOICE_HASH: "invoice_hash",
        tlv.Tag.SIGNATURE: "signature",
        tlv.Tag.PUBLIC_KEY: "public_key",
        tlv.Tag.CSID_SIGNATURE: "csid_signature",
    }
    out = {}
    for tag, value in raw.items():
        key = names.get(tag, f"tag_{tag}")
        if tag in (tlv.Tag.PUBLIC_KEY,):
            out[key] = f"<{len(value)} bytes>"
        else:
            try:
                out[key] = value.decode("utf-8")
            except UnicodeDecodeError:
                out[key] = f"<{len(value)} bytes>"
    return out

File: app/lib/zatca/qr.dart
/// Assemble the ZATCA invoice QR payload.
///
/// Dart port of `zatca/zatca/qr.py`. This is what gets printed on the
/// customer's receipt. It is produced on the device, at the moment of sale,
/// with no network — which is the whole reason each tablet holds its own
/// CSID.
///
/// Money arrives as integer halalas (as it is stored everywhere else) and is
/// formatted to two decimals only at this boundary, because ZATCA wants a
/// decimal string. The conversion is integer division, never float
/// arithmetic.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'signing.dart';
import 'tlv.dart';

/// 1234 -> '12.34'. Integer maths only; no float ever touches a total.
String halalasToDecimalString(int halalas) {
  if (halalas < 0) {
    throw ArgumentError('amount cannot be negative');
  }
  return '${halalas ~/ 100}.${(halalas % 100).toString().padLeft(2, '0')}';
}

/// ISO 8601 in UTC with a trailing Z, no microseconds.
String zatcaTimestamp(DateTime when) {
  final utc = when.toUtc();
  return '${utc.year.toString().padLeft(4, '0')}-'
      '${utc.month.toString().padLeft(2, '0')}-'
      '${utc.day.toString().padLeft(2, '0')}T'
      '${utc.hour.toString().padLeft(2, '0')}:'
      '${utc.minute.toString().padLeft(2, '0')}:'
      '${utc.second.toString().padLeft(2, '0')}Z';
}

class QrInput {
  const QrInput({
    required this.sellerName,
    required this.vatNumber,
    required this.issuedAt,
    required this.totalWithVat,
    required this.vatTotal,
    required this.invoiceHash,
    required this.publicKeyDer,
    required this.csidSignature,
  });

  /// Arabic name as registered with ZATCA.
  final String sellerName;

  /// 15 digits.
  final String vatNumber;
  final DateTime issuedAt;

  /// halalas
  final int totalWithVat;

  /// halalas
  final int vatTotal;

  /// base64
  final String invoiceHash;
  final Uint8List publicKeyDer;

  /// base64, ZATCA's signature over the public key.
  final String csidSignature;
}

/// Return the base64 TLV payload to encode as a QR on the receipt.
String buildQr(QrInput data, String privatePem) {
  if (data.vatNumber.length != 15 ||
      !RegExp(r'^\d{15}$').hasMatch(data.vatNumber)) {
    throw ArgumentError('VAT registration number must be 15 digits');
  }
  if (data.vatTotal > data.totalWithVat) {
    throw ArgumentError('VAT cannot exceed the invoice total');
  }

  // The stamp is over the invoice hash, which is what binds the QR to the
  // document it was printed for.
  final signature = signDigest(privatePem, ascii.encode(data.invoiceHash));

  return encodeTlvBase64({
    ZatcaTag.sellerName: data.sellerName,
    ZatcaTag.vatNumber: data.vatNumber,
    ZatcaTag.timestamp: zatcaTimestamp(data.issuedAt),
    ZatcaTag.invoiceTotal: halalasToDecimalString(data.totalWithVat),
    ZatcaTag.vatTotal: halalasToDecimalString(data.vatTotal),
    ZatcaTag.invoiceHash: data.invoiceHash,
    ZatcaTag.signature: signature,
    ZatcaTag.publicKey: data.publicKeyDer,
    ZatcaTag.csidSignature: data.csidSignature,
  });
}

/// Decode a QR payload for support and debugging.
Map<String, String> inspectQr(String qrBase64) {
  const names = {
    ZatcaTag.sellerName: 'seller_name',
    ZatcaTag.vatNumber: 'vat_number',
    ZatcaTag.timestamp: 'timestamp',
    ZatcaTag.invoiceTotal: 'total_with_vat',
    ZatcaTag.vatTotal: 'vat_total',
    ZatcaTag.invoiceHash: 'invoice_hash',
    ZatcaTag.signature: 'signature',
    ZatcaTag.publicKey: 'public_key',
    ZatcaTag.csidSignature: 'csid_signature',
  };
  final out = <String, String>{};
  decodeTlvBase64(qrBase64).forEach((tag, value) {
    final key = names[tag] ?? 'tag_$tag';
    if (tag == ZatcaTag.publicKey) {
      out[key] = '<${value.length} bytes>';
    } else {
      try {
        out[key] = utf8.decode(value);
      } on FormatException {
        out[key] = '<${value.length} bytes>';
      }
    }
  });
  return out;
}
